Added throttling capabilities for high-frequency tf logging

fairlight1337 opened the pull request:
Added tf logging throttling capabilities originally introduced in https://github.com/code-iai/ros-mongodb_log. A transform is only logged when either this transform has not been logged before, or when the new version of this transform is sufficiently different from the one logged before. Additional command line parameters can be used to control how throttling is done:
- `-k <d>`: Cartesian (vectorial) distance (in meters) threshold between the old and the new transform
- `-l <d>`: Angular diastance (in deg) threshold between the old and the new transform
- `-g <d>`: At least log every transform every `d` seconds, even if nothing changed
- `-a`: Always log, don't throttle

fairlight1337 committed:
Added throttling capabilities for high-frequency tf logging

Added tf logging throttling capabilities originally introduced in https://github.com/code-iai/ros-mongodb_log. A transform is only logged when either this transform has not been logged before, or when the new version of this transform is sufficiently different from the one logged before. Additional command line parameters can be used to control how throttling is done:

 * `-k <d>`: Cartesian (vectorial) distance (in meters) threshold between the old and the new transform
 * `-l <d>`: Angular diastance (in deg) threshold between the old and the new transform
 * `-g <d>`: At least log every transform every `d` seconds, even if nothing changed
 * `-a`: Always log, don't throttle

strands-jenkins commented:
Can one of the STRANDS admins verify this patch? Post a comment containing "ok to test" to enable Jenkins builds for this pull request.

fairlight1337 commented:
Hey,

I saw that you guys also use the mongodb logger. We found the throttling feature immensely useful, as it helps to not clutter the database with data that isn't changing much (or at all). Hope this helps you in some way as well.

marc-hanheide commented:
ok to test

marc-hanheide commented:
Thanks @fairlight1337 very useful. We'll have a look.

marc-hanheide commented on mongodb_log/src/mongodb_log_tf.cpp:
```diff
@@ -44,84 +63,159 @@ static pthread_mutex_t out_counter_mutex = PTHREAD_MUTEX_INITIALIZER;
-  while ((c = getopt(argc, argv, "t:m:n:c:")) != -1) {
+  while ((c = getopt(argc, argv, "t:m:n:c:ak:l:g:")) != -1) {
     if ((c == '?') || (c == ':')) {
       printf("Usage: %s -t topic -m mongodb -n nodename -c collection\n", argv[0]);
```
The usage message is now not complete any longer, @fairlight1337. Do you want to add the new options, please?

fairlight1337 replied:
Okay, good point. I'll add the new options once I changed the default behaviour (see my other commit).

fairlight1337 replied:
Addressed in https://github.com/fairlight1337/mongodb_store/commit/2a565050ab4c8285d687b2e280d44617cd4f1410.

marc-hanheide commented on mongodb_log/src/mongodb_log_tf.cpp:
```diff
@@ -133,9 +227,17 @@ main(int argc, char **argv)
       nodename = optarg;
     } else if (c == 'c') {
       collection = optarg;
+    } else if (c == 'a') {
```
This now changes the default behaviour. In order to get the "old" behaviour one needs to give a new flag `-a`. This should be avoided IMO, as behaviour should be kept the same when additional functionality is it. Do you agree, @fairlight1337 ?

fairlight1337 replied:
Totally agree. I'll change the default behaviour.

fairlight1337 replied:
Addressed in https://github.com/fairlight1337/mongodb_store/commit/a8c5a6697bf25c8d9f1d76389c24b6522f21ff34.

fairlight1337 committed:
Changed default behaviour back to its former way

The 'a' command line parameter now activates throttling; not specifying it makes the logger log all tf transformations.

fairlight1337 committed:
Extended usage output string by new command line options

hawesie commented:
Cool, thanks. I'm still looking for a little bit of spare time to run this myself, so sorry for the delay.

marc-hanheide commented:
@hawesie did you have a look at this?

strands-jenkins commented:
Can one of the STRANDS admins verify this patch? Post a comment containing "ok to test" to enable Jenkins builds for this pull request.

hawesie commented:
Not yet :(

I will try to get these in and released by next week.

hawesie commented:
ok to test

hawesie commented:
Tested ok! The only issue is that the type-specific logger is not launched automatically from the mongodb_log.py for some reason, but I will look at that separately.
